nickel/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+//! A stable library interface to the Nickel language.
+//!
+//! Most of the Nickel language is implemented in `nickel-lang-core`, but
+//! `nickel-lang-core` exposes implementation details that frequently change.
+//! This library implements a more basic interface, with fewer features. Its
+//! goals are:
+//!
+//! - Simplicity: it should be easy to do basic things, like evaluating
+//!   some Nickel code into a structured representation.
+//! - Stability: we aim not to break public APIs. This doesn't mean
+//!   we'll *never* break them, just that stability is a priority.
+//! - Embeddability: our interfaces are designed to be usable from
+//!   other languages.
+
+use std::{
+    cell::RefCell,
+    ffi::OsString,
+    io::{Cursor, Write},
+    rc::Rc,
+};
+
+use codespan_reporting::term::termcolor::{Ansi, NoColor, WriteColor};
+use malachite::base::{num::conversion::traits::RoundingFrom, rounding_modes::RoundingMode};
+
+use nickel_lang_core::{
+    error::{report::DiagnosticsWrapper, IntoDiagnostics, NullReporter},
+    eval::{cache::CacheImpl, Closure, Environment},
+    files::Files,
+    identifier::Ident,
+    program::Program,
+    serialize::{to_string, validate, ExportFormat},
+    term::{self, record::RecordData, RichTerm, Term},
+};
+
+pub use nickel_lang_core::error::report::ErrorFormat;
+
+/// Provides a destination for the output of `std.trace`.
+#[derive(Clone)]
+pub struct Trace {
+    // This is a little annoying, in that VirtualMachine already takes a generic
+    // trace and boxes it. Since we don't want a generic parameter on Context,
+    // it means we need to double-box it.
+    inner: Rc<RefCell<dyn Write>>,
```

Comment:
I suppose one solution would be to pass a `Box<dyn Write>` to the VM, instead of having it do the boxing itself. However here you use `Rc` instead. Is there a reason why we would want the context to be cloneable (other than "it might be a good idea, UX-wise, in general")? Another solution (in addition to the change to `VirtualMachine` and `Program` to NOT do the boxing themselves) would be to separate the trace from the rest of the context, even if it makes the whole idea of having only one big context around a tad less appealing.